File: examplepatches/Wake-Old-Hyphen/hushfeed/extensions/tiktok/src/main/java/app/morphe/extension/tiktok/interaction/RailHold.java
/*
 * Copyright 2026 Hushfeed contributors
 * https://github.com/SysAdminDoc/hushfeed
 */
package app.morphe.extension.tiktok.interaction;

import android.os.Handler;
import android.os.SystemClock;
import android.view.View;
import android.view.ViewConfiguration;

import app.morphe.extension.shared.diagnostics.HookStatus;
import app.morphe.extension.shared.settings.BooleanSetting;
import app.morphe.extension.tiktok.settings.Settings;

/**
 * A long press on the Comment, Share or Favorites button plays the video at the hold speed while
 * it's held (upstream #87), the way a long press did before TikTok gave those buttons menus of
 * their own.
 *
 * <p>TikTok's hold gesture sits on the feed cell, sees every touch and speeds a video up when a
 * press lands near the left or right edge of the screen, where the side buttons are. As the press
 * lands it asks the buttons whether the press is on one of them, and a press on one is not a
 * hold. So the patch answers "not on it" for a chosen button, and TikTok's own hold runs there:
 * the hold speed, its banner, the pull-down lock and the release. The button's own long press,
 * the emoji row on Comment, the quick share row on Share and the new collection offer on
 * Favorites, is skipped for that press, so a hold doesn't open a menu too, and letting go doesn't
 * count as a tap. Comment's and Favorites' are skipped in the method that opens the menu, and
 * Share's by wrapping the listeners TikTok sets.
 *
 * <p>Only a press the hold check let through skips the menu, and only while TikTok still means to
 * hold it. The check turns a press down before it asks the buttons when Hushfeed has the edge for
 * a seek or the long press for an action of its own, and when TikTok won't speed up at all. TikTok
 * drops a hold it hasn't started yet when the finger moves off the spot or a second finger lands,
 * and ends every press on release, and each of those forgets the pass. The menu then opens as it
 * always did, as long as the button's own timer hasn't fired yet. Favorites' fires at 200 ms,
 * before the hold starts at 300 ms, so a release between the two gets neither.
 */
public final class RailHold {
    static final String FAMILY = "rail hold";

    /**
     * How long past the long-press timeout a long press still belongs to the press the hold check
     * let through. The check runs as the press lands and the long press comes a timeout later.
     */
    static final long GRACE_MS = 1_000;

    /** A side button a hold can be chosen for, and when the check last let a press on it through. */
    enum Button {
        COMMENT("comment", Settings.RAIL_HOLD_COMMENT),
        SHARE("share", Settings.RAIL_HOLD_SHARE),
        FAVORITES("favorites", Settings.RAIL_HOLD_FAVORITES);

        final String label;
        final BooleanSetting setting;
        /** In uptime, or -1. Touched on the main thread only, where TikTok's touches run. */
        long heldAt = -1;

        Button(String label, BooleanSetting setting) {
            this.label = label;
            this.setting = setting;
        }

        /** TikTok's hold check asks whether a press is on this button, where a hold doesn't start. */
        boolean stopsHold(boolean onButton) {
            HookStatus.bound(FAMILY, label + " check");
            if (!onButton || !setting.get()) return onButton;
            HookStatus.bound(FAMILY, "held on " + label);
            heldAt = SystemClock.uptimeMillis();
            return false;
        }

        /** Whether the long press on this button now is one the hold check just let through. */
        boolean heldJustNow() {
            return heldAt >= 0 && setting.get()
                    && SystemClock.uptimeMillis() - heldAt <= ViewConfiguration.getLongPressTimeout() + GRACE_MS;
        }

        /** TikTok's long press on this button, about to open its menu. True skips the menu. */
        boolean skipsMenu() {
            HookStatus.bound(FAMILY, label + " long press");
            if (!heldJustNow()) return false;
            HookStatus.bound(FAMILY, label + " menu skipped");
            return true;
        }
    }

    private RailHold() {
    }

    public static boolean commentStopsHold(boolean onComment) {
        return Button.COMMENT.stopsHold(onComment);
    }

    public static boolean shareStopsHold(boolean onShare) {
        return Button.SHARE.stopsHold(onShare);
    }

    public static boolean favoritesStopsHold(boolean onFavorites) {
        return Button.FAVORITES.stopsHold(onFavorites);
    }

    /**
     * Comment's long press, about to open the emoji row. TikTok marks the press as a long one
     * before it gets here, so a true answer skips the row and letting go still isn't a tap.
     */
    public static boolean skipCommentMenu() {
        return Button.COMMENT.skipsMenu();
    }

    /**
     * Favorites' long press, about to offer a new collection. TikTok marks the press as a long one
     * before it gets here, so a true answer skips the offer and letting go doesn't save the video.
     */
    public static boolean skipFavoritesMenu() {
        return Button.FAVORITES.skipsMenu();
    }

    /** Where TikTok gives the Share button, or one of its children, its long press. */
    public static void setShareLongClick(View button, View.OnLongClickListener listener) {
        HookStatus.bound(FAMILY, "share listener");
        button.setOnLongClickListener(listener == null ? null : new ShareMenu(listener));
    }

    /**
     * TikTok's hold, dropping the hold it had set to start 300 ms after the press landed: the
     * finger moved off the spot, a second finger landed, or the press ended. Every pass goes with
     * it, since none of them belongs to a press TikTok will hold.
     */
    public static void holdDropped(Handler handler, Runnable pending) {
        handler.removeCallbacks(pending);
        HookStatus.bound(FAMILY, "hold dropped");
        forgetPasses();
    }

    private static void forgetPasses() {
        for (Button button : Button.values()) button.heldAt = -1;
    }

    static void resetForTests() {
        forgetPasses();
    }

    /**
     * Share's own long press, which opens its menu unless the press is being held at the hold
     * speed. Asked each time, so the switch applies to buttons already on screen. Taking the long
     * press is what keeps letting go from counting as a tap.
     */
    static final class ShareMenu implements View.OnLongClickListener {
        final View.OnLongClickListener tikTok;

        ShareMenu(View.OnLongClickListener tikTok) {
            this.tikTok = tikTok;
        }

        @Override public boolean onLongClick(View view) {
            return Button.SHARE.skipsMenu() || tikTok.onLongClick(view);
        }
    }
}

File: examplepatches/SysAdminDoc/hushfeed/patches/src/main/kotlin/app/morphe/patches/tiktok/interaction/gesture/RailHold.kt
/*
 * Copyright 2026 Hushfeed contributors
 * https://github.com/SysAdminDoc/hushfeed
 */
package app.morphe.patches.tiktok.interaction.gesture

import app.morphe.patcher.Fingerprint
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.extensions.InstructionExtensions.replaceInstruction
import app.morphe.patcher.patch.PatchException
import app.morphe.patcher.util.proxy.mutableTypes.MutableMethod
import app.morphe.patcher.util.smali.ExternalLabel
import app.morphe.patches.tiktok.shared.guardAtEntry
import app.morphe.patches.tiktok.shared.requireLocals
import app.morphe.util.getReference
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.Method
import com.android.tools.smali.dexlib2.iface.instruction.FiveRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.Instruction
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.reference.FieldReference
import com.android.tools.smali.dexlib2.iface.reference.MethodReference

internal const val RAIL_HOLD_EXTENSION = "Lapp/morphe/extension/tiktok/interaction/RailHold;"
internal const val VIDEO_COMMENT_ASSEM = "Lcom/ss/android/ugc/aweme/feed/assem/videocomment/VideoCommentAssem;"
internal const val VIDEO_SHARE_ASSEM = "Lcom/ss/android/ugc/aweme/feed/assem/share/VideoShareAssem;"
internal const val COMMENT_ABILITY = "Lcom/ss/android/ugc/aweme/feed/assem/ability/IVideoCommentAbility;"
internal const val SHARE_ABILITY = "Lcom/ss/android/ugc/aweme/feed/assem/ability/IVideoShareAbility;"
internal const val FAVORITES_ABILITY = "Lcom/ss/android/ugc/feed/platform/cell/ability/VideoFavoriteAssemAbility;"
internal const val COMMENT_MENU_KEY = "video_comment_button_long_click"
internal const val FAVORITES_MENU_KEY = "video_favorite_button_long_click"
internal const val FAVORITES_MENU_EVENT = "long_press_favorite"
private const val VIEW = "Landroid/view/View;"
private const val ON_LONG_CLICK = "Landroid/view/View\$OnLongClickListener;"

/** Whether the instruction is TikTok's own View.setOnLongClickListener call. */
internal fun Instruction.setsLongClick(): Boolean =
    opcode == Opcode.INVOKE_VIRTUAL && getReference<MethodReference>()?.let {
        it.definingClass == VIEW && it.name == "setOnLongClickListener" &&
            it.parameterTypes.map(CharSequence::toString) == listOf(ON_LONG_CLICK)
    } == true

/**
 * The Comment button's long press, which opens the emoji row. TikTok reaches it two ways: a
 * long-click listener on the button, or, when a server flag says so, the button's touch listener
 * after a timer of its own that also keeps letting go from counting as a tap. On 47.0.3 the S22
 * gets the second, where no long-click listener is set at all. Both call this one method, which
 * hands the button's place to the emoji row under a key naming it, and that key picks the method
 * out of the class on every build.
 */
internal object CommentMenuFingerprint : Fingerprint(
    definingClass = VIDEO_COMMENT_ASSEM,
    returnType = "V",
    parameters = listOf(),
    strings = listOf(COMMENT_MENU_KEY),
)

/**
 * The Favorites button's long press, which offers a new collection. TikTok's touch listener on the
 * button runs it from a timer that first marks the press as a long one, so letting go doesn't save
 * the video. It's a lambda with an obfuscated name. What picks it out is that it both logs the
 * long press and hands the button's place to the offer under a key naming it, and only it does.
 */
internal object FavoritesMenuFingerprint : Fingerprint(
    name = "invoke",
    returnType = "Ljava/lang/Object;",
    parameters = listOf(),
    strings = listOf(FAVORITES_MENU_KEY, FAVORITES_MENU_EVENT),
)

/** The Share button's view setup, which gives it and two of its children their long presses. */
internal object ShareViewCreatedFingerprint : Fingerprint(
    definingClass = VIDEO_SHARE_ASSEM,
    name = "onViewCreated",
    returnType = "V",
    parameters = listOf(VIEW),
    custom = { method, _ -> method.implementation?.instructions?.any { it.setsLongClick() } == true },
)

/**
 * The indexes in TikTok's hold check of the asks to [ability] whether a press is on its button:
 * an interface call taking the press's two floats and answering Z, its answer moved straight
 * into a register. The index is the move's.
 */
internal fun Method.railHitTests(ability: String): List<Int> {
    val instructions = implementation?.instructions?.toList() ?: return emptyList()
    return instructions.indices.filter { index ->
        val instruction = instructions[index]
        instruction.opcode == Opcode.INVOKE_INTERFACE && instruction.getReference<MethodReference>()?.let {
            it.definingClass == ability && it.returnType == "Z" &&
                it.parameterTypes.map(CharSequence::toString) == listOf("F", "F")
        } == true && instructions.getOrNull(index + 1)?.opcode == Opcode.MOVE_RESULT
    }.map { it + 1 }
}

/**
 * Hands each Comment, Share and Favorites ask in TikTok's hold check to the extension, which
 * answers "not on it" for a button a hold plays at the hold speed on. Inserted after each answer's
 * move, last one first so the indexes found hold.
 */
internal fun MutableMethod.answerRailHitTests() {
    val asks = railHitTests(COMMENT_ABILITY).map { it to "commentStopsHold" } +
        railHitTests(SHARE_ABILITY).map { it to "shareStopsHold" } +
        railHitTests(FAVORITES_ABILITY).map { it to "favoritesStopsHold" }
    for (answer in listOf("commentStopsHold", "shareStopsHold", "favoritesStopsHold")) {
        if (asks.none { it.second == answer }) {
            throw PatchException("Long-press controls: the hold check no longer asks the button $answer answers for.")
        }
    }
    asks.sortedByDescending { it.first }.forEach { (moveAt, answer) ->
        val register = (implementation!!.instructions[moveAt] as OneRegisterInstruction).registerA
        addInstructions(
            moveAt + 1,
            """
                invoke-static/range {v$register .. v$register}, $RAIL_HOLD_EXTENSION->$answer(Z)Z
                move-result v$register
            """,
        )
    }
}

/** Returns from the Comment button's long press before the emoji row when the press is held. */
internal fun MutableMethod.skipCommentMenuWhenHeld() =
    guardAtEntry("Long-press controls", "invoke-static {}, $RAIL_HOLD_EXTENSION->skipCommentMenu()Z", "return-void")

/**
 * Leaves the Favorites long press before it logs or offers anything when the press is held. The
 * lambda answers Kotlin's Unit, whose field name is obfuscated too, so the skip jumps to where
 * the lambda itself loads Unit and returns it.
 */
internal fun MutableMethod.skipFavoritesMenuWhenHeld() {
    val instructions = implementation!!.instructions.toList()
    val unit = instructions.indices.lastOrNull { index ->
        instructions[index].opcode == Opcode.SGET_OBJECT &&
            instructions[index].getReference<FieldReference>()?.type == "Lkotlin/Unit;" &&
            instructions.getOrNull(index + 1)?.opcode == Opcode.RETURN_OBJECT &&
            (instructions[index] as OneRegisterInstruction).registerA ==
            (instructions[index + 1] as OneRegisterInstruction).registerA
    } ?: throw PatchException("Long-press controls: $definingClass->$name no longer returns Unit at its end.")
    requireLocals("Long-press controls", 1)
    addInstructionsWithLabels(
        0,
        """
            invoke-static {}, $RAIL_HOLD_EXTENSION->skipFavoritesMenu()Z
            move-result v0
            if-nez v0, :rail_hold_unit
        """,
        ExternalLabel("rail_hold_unit", getInstruction(unit)),
    )
}

/** Sends each of the method's long-click sets through [hook], which wraps TikTok's listener. */
internal fun MutableMethod.routeLongClicks(hook: String) {
    val sites = implementation!!.instructions.withIndex().filter { it.value.setsLongClick() }.map { it.index }
    if (sites.isEmpty()) throw PatchException("Long-press controls: $definingClass->$name sets no long press.")
    sites.forEach { index ->
        val call = implementation!!.instructions[index] as FiveRegisterInstruction
        replaceInstruction(
            index,
            "invoke-static {v${call.registerC}, v${call.registerD}}, $RAIL_HOLD_EXTENSION->$hook($VIEW$ON_LONG_CLICK)V",
        )
    }
}
